### PR TITLE
Hello angular motoko

### PR DESCRIPTION
**Pull request by shawndotey**
**Overview**
This is an example of serving and deploying the Internet Computer's motoko project along side Angular.

**Requirements**
The project will:

- Serve a development environment with hot reloading, both Angular and Motoko.
- Deploy it's build to live Canisters.

**Considerations**
- The npm `build`, `prebuild` and `prestart` commands are used by the IC processes. Using them may have unintended side effects with other build processes.
- Currently, the Typescript generated by DFX has trailing commas, which produces a conflict with Angular's default build configurations. Fixing this is beyound the scope of this example.
- This example is intended to help those already familar with Angular to get started with the IC. This example is designed to be as straight forward as possible but it is not intended as a tool for learning Angular.

### CODE REVIEW

File: motoko/hello-angular-motoko/README.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# hello-angular-motoko
+This is an example of locally implementing and deploying the Internet Computer's motoko project (0.10.1) along side Angular 14.
```
Comment (sesi200) [inline motoko/hello-angular-motoko/README.md:2]:
```suggestion
This is an example of a Motoko project with an Angular 14 frontend.
```

File: motoko/hello-angular-motoko/README.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+-----------
+## Getting started
+
+- [Install DFX](https://sdk.dfinity.org/docs/quickstart/local-quickstart.html). Please keep in mind the dfx cli currently only runs on Linux and Apple based PCs.
```
Comment (sesi200) [inline motoko/hello-angular-motoko/README.md:7]:
```suggestion
- [Install DFX](https://internetcomputer.org/docs/current/developer-docs/quickstart/local-quickstart).
```
Updating with the current link, removing system requirements because that is a future stale bit of information.

File: motoko/hello-angular-motoko/README.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+- You can use the following command:
+
+    ```bash
+    # Starts the replica, running in the background
+    # Serves Angular and dfx canisters locally with hot reloading
+    npm run start
+    ```
+    Once the job completes, your application will be available at `http://localhost:4200`.
+
+    If you make any changes while in development mode, the project will recompile on both the Angular and motoko side.
```
Comment (sesi200) [inline motoko/hello-angular-motoko/README.md:32]:
```suggestion
    If you make any changes while in development mode, the project will recompile on both the Angular and Motoko side.
```

File: motoko/hello-angular-motoko/README.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+|`restart:clean:dfx`|Starts the local canister execution environment and web server processes in a clean state by removing checkpoints from your project cache. You can use this to set your project cache to a new state when troubleshooting or debugging.|
+|`generate:motoko`| Generate canister type declarations for motoko supported programming languages. This will generate the `src/declarations` folder and remove conflicts. These generated js files allow you to consume the motoko language service on the front end. i.e `const  ic_service = require('src/declarations/motoko').motoko;`
+
+## Notes
+- The npm `build`, `prebuild` and `prestart` commands are used by the IC processes. Using them may have unintended side effects with other build processes.
+- Currently, the Typescript generated by DFX has trailing commas, which produces a conflict with Angular's default build configurations. Fixing this is beyound the scope of this example.
+- This example is intended to help those already familar with Angular to get started with the IC. This example is designed to be as straight forward as possible but it is not intended as a tool for learning Angular.
```
Comment (sesi200) [inline motoko/hello-angular-motoko/README.md:71]: This should not link out to parts of external tutorials. The readme should be self-contained. Can you please simplify this a bit?

Side note: What does not work with `dfx deploy --network ic`? All projects we have so far work with a simple `dfx deploy`.

File: motoko/hello-angular-motoko/README.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+-----
+
+### To learn more about working with dfx, see the following documentation available online:
+
+- [Quick Start](https://sdk.dfinity.org/docs/quickstart/quickstart-intro.html)
+- [SDK Developer Tools](https://sdk.dfinity.org/docs/developers-guide/sdk-guide.html)
+- [Motoko Programming Language Guide](https://sdk.dfinity.org/docs/language-guide/motoko.html)
+- [Motoko Language Quick Reference](https://sdk.dfinity.org/docs/language-guide/language-manual.html)
+- [JavaScript API Reference](https://erxue-5aaaa-aaaab-qaagq-cai.raw.ic0.app)
```
Comment (sesi200) [inline motoko/hello-angular-motoko/README.md:81]: Most of this is not dfx-specific (as line 75 notes).

File: motoko/hello-angular-motoko/canister_ids.json [hunk 6]
```diff
@@ -0,0 +1,8 @@
+{
+  "motoko": {
+    "ic":"65yca-4yaaa-aaaak-qaqcq-cai"
+  },
+  "www": {
+    "ic": "zcp5t-tqaaa-aaaaj-aijea-cai"
+  }
+}
```
Comment (sesi200) [inline motoko/hello-angular-motoko/canister_ids.json:8]: This is your own `ic` deployment. I don't think you want those to be in the examples repo. You're welcome to link to them in the readme as a live example, however.

File: motoko/hello-angular-motoko/karma.conf.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
+// Karma configuration file, see link for more information
+// https://karma-runner.github.io/1.0/config/configuration-file.html
```
Comment (sesi200) [inline motoko/hello-angular-motoko/karma.conf.js:2]: This is a testing framework, right? Can we add 1-2 tests that the project still works as part of CI? This would make it much easier to maintain in the future (e.g. updating dependencies).

File: motoko/hello-angular-motoko/package.json [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    "print:dfx:www": "echo local IC server: http://127.0.0.1:8000/?canisterId=$(dfx canister id www)",
+    "print:dfx:motoko": "echo local IC server: http://127.0.0.1:8000/?canisterId=$(dfx canister id motoko)",
+    "generate:motoko": "dfx generate motoko",
+    "restart:dfx": "dfx stop && dfx start --background",
+    "restart:clean:dfx": "dfx stop && dfx start --background --clean",
+    "refresh:motoko": "dfx deploy motoko && npm run copy:dfx",
+    "copy:dfx": "rsync -avr .dfx/$(echo ${DFX_NETWORK:-'**'})/canisters/** --exclude='assets/' --exclude='idl/' --exclude='*.wasm' --exclude='*.most' --exclude='*.ts' --delete src/declarations",
+    "build:prod": "export NODE_ENV=production && ng build --configuration production",
+    "build:dev": "export NODE_ENV=development && ng build --configuration development"
+  },
```
Comment (sesi200) [inline motoko/hello-angular-motoko/package.json:35]: I think this is nice to keep as an 'advanced' part of the example, but it should not be the only way to deploy/control the project.

**Review by sesi200 — CHANGES_REQUESTED** [review]
Hi Shawn, thanks a lot for your PR, I can see this was a lot of work! I think this would be a great addition to our examples.

There's a few things to clean up, especially regarding documentation, but I look forward to merging this.

File: motoko/hello-angular-motoko/README.md (hunk 4 above)
Comment (shawndotey) [reply motoko/hello-angular-motoko/README.md]: Updated with self-contained Prerequesites. 
Side Note - `dfx deploy --network ic --no-wallet` deploys to the IC (https://ic0.app).  Without it, it deploys locally.

File: motoko/hello-angular-motoko/package.json (hunk 8 above)
Comment (shawndotey) [reply motoko/hello-angular-motoko/package.json]: Ok, I think I may need to circle back on that one.

File: motoko/hello-angular-motoko/karma.conf.js (hunk 7 above)
Comment (shawndotey) [reply motoko/hello-angular-motoko/karma.conf.js:2]: Added a Test Suite for the motoko service.

File: motoko/hello-angular-motoko/canister_ids.json (hunk 6 above)
Comment (shawndotey) [reply motoko/hello-angular-motoko/canister_ids.json]: Done.

File: motoko/hello-angular-motoko/README.md (hunk 5 above)
Comment (shawndotey) [reply motoko/hello-angular-motoko/README.md]: Fixed.

**Comment by shawndotey** [comment]
@sesi200 That should address all the key issues except not using npm as the primary implementation method. Please let me know if there is any other blockers. Thank you!

File: motoko/hello-angular-motoko/README.md [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        "motoko": {
+            # replace this with your unique cansiter from the above guide
+            "ic":"your-unique-canister-goes-here"
+        },
+        "www": {
+            # replace this with a second unique cansiter from the above guide
+            "ic": "another-unique-canister-goes-here"
+        }
+    
+        ```
```
Comment (sesi200) [inline motoko/hello-angular-motoko/README.md:45]:
```suggestion
- For live IC deployment: A wallet set up as described in the [quickstart guide](https://internetcomputer.org/docs/current/developer-docs/quickstart/hello10mins/).
```
With a configured wallet (which I think is a reasonable demand), this could all be replaced by `dfx canister --network ic create --all --with-cycles 1000000000000`, which `dfx deploy --network ic` automatically does when you also add `--with-cycles 1000000000000`.

File: motoko/hello-angular-motoko/README.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+## Testing Locally:
+- From the project root.
+
+    ```bash
+    npm run test
+    ```
+
+
+-----
```
Comment (sesi200) [inline motoko/hello-angular-motoko/README.md:86]: Can you also mention the normal dfx commands for this?

File: motoko/hello-angular-motoko/canister_ids.json [hunk 11]
```diff
@@ -0,0 +1,8 @@
+{
+  "motoko": {
+    "ic":"your-unique-canister-goes-here"
+  },
+  "www": {
+    "ic": "another-unique-canister-goes-here"
+  }
+}
```
Comment (sesi200) [inline motoko/hello-angular-motoko/canister_ids.json:8]: When you remove the above (quite convoluted) way of creating canisters, this file can be deleted.

File: motoko/hello-angular-motoko/package.json [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    "test": "npm run restart:dfx && ng test",
+    "deploy:dfx": "npm run restart:dfx && npm run build:prod && dfx deploy --network ic --no-wallet",
+    "serve:both": "concurrently -k \"npm run watch:motoko\" \"ng serve && dfx stop\"",
+    "watch:motoko": "nodemon --watch src --ext mo -x \"npm run refresh:motoko\"",
+    "print:dfx:www": "echo local IC server: http://127.0.0.1:8000/?canisterId=$(dfx canister id www)",
+    "print:dfx:motoko": "echo local IC server: http://127.0.0.1:8000/?canisterId=$(dfx canister id motoko)",
+    "generate:motoko": "dfx generate motoko",
+    "restart:dfx": "dfx stop && dfx start --background",
+    "restart:clean:dfx": "dfx stop && dfx start --background --clean",
+    "refresh:motoko": "dfx deploy motoko && npm run copy:dfx",
```
Comment (sesi200) [inline motoko/hello-angular-motoko/package.json:32]: why is `copy:dfx` included here? And why _after_ the deploy? I don't understand what the copy step does here.

File: motoko/hello-angular-motoko/package.json [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    "copy:dfx": "rsync -avr .dfx/$(echo ${DFX_NETWORK:-'**'})/canisters/** --exclude='assets/' --exclude='idl/' --exclude='*.wasm' --exclude='*.most' --exclude='*.ts' --delete src/declarations",
+    "build:prod": "export NODE_ENV=production && ng build --configuration production",
+    "build:dev": "export NODE_ENV=development && ng build --configuration development"
```
Comment (sesi200) [inline motoko/hello-angular-motoko/package.json:35]: Does this mean the normal `dfx build --network <ic/local>` will not work for this project?

**Comment by sesi200** [comment]
@shawndotey sorry for the comment spam. I think you're addressing my concerns very well. Some of my confusion can also stem from me not being a frontend guy at all. It's totally fine to answer some of my points with "that's just how frontend/angular works"

File: motoko/hello-angular-motoko/package.json [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    "Canister"
+  ],
+  "scripts": {
+    "build": "",
+    "prebuild": "",
+    "prestart": "",
+    "init:dfx": "npm run restart:clean:dfx && npm run refresh:motoko && dfx stop",
+    "start": "npm run restart:dfx && npm run refresh:motoko && npm run serve:both && npm run print:ng",
+    "test": "npm run restart:dfx && ng test",
+    "deploy:dfx": "npm run restart:dfx && npm run build:prod && dfx deploy --network ic --no-wallet",
```
[the remark this answers is not in the document]
Comment (shawndotey) [reply motoko/hello-angular-motoko/package.json:24]: - `restart:dfx` allows consistent behavior while deploying or starting am instance. Sometimes there are conflicts with a prior started dfx canisters that results in needing to either manually stopping processes or restating your computer.
- `build:prod` is different than building dfx canisters. `build:prod` builds the front end Angular framework, not the canisters. I think I can rework the `dfx deploy` to use the `build`, `prebuild`,`prestart`  steps to run the `build:prod` command but I encountered some conflicts when I last tried. For now I know this works.

File: motoko/hello-angular-motoko/package.json (hunk 12 above)
Comment (shawndotey) [reply motoko/hello-angular-motoko/package.json:32]: `refresh:motoko` is used when initializing and running the project in development. The `copy:dfx` is the same as the `copy:types` command generated by a default new project (`dfx new <project_name> --frontend), except it removes typescript files as they currently have an issue with trailing commas. This command is used to refresh the front end canister declarations after a hot reload and on the project initialization. Otherwise, they are not available for the the front end to consume.

File: motoko/hello-angular-motoko/README.md (hunk 9 above)
Comment (shawndotey) [reply motoko/hello-angular-motoko/README.md:45]: @sesi200 
- How does the project know what canisters to deploy to? This is the recommended method as given by @krpeacock. I imagine it will do well for now.
- As you stated before, simply providing a link without explicitly describing the instructions is bad practice.

File: motoko/hello-angular-motoko/README.md (hunk 10 above)
Comment (shawndotey) [reply motoko/hello-angular-motoko/README.md:86]: Yes, can we make do with this for now as it is inferred in the actual command(s)?

Comment (shawndotey) [reply motoko/hello-angular-motoko/README.md:86]: @sesi200

File: motoko/hello-angular-motoko/karma.conf.js (hunk 7 above)
Comment (shawndotey) [reply motoko/hello-angular-motoko/karma.conf.js:2]: @sesi200, Yes. This file just needs to be at the project root with 
1) `motoko-classes-example-...` replaced with this examples corresponding name?
2) The `run:` commands replaced with the appropriate commands?

File: motoko/hello-angular-motoko/package.json (hunk 14 above)
Comment (sesi200) [inline motoko/hello-angular-motoko/package.json:24]:
```suggestion
    "deploy:dfx": "npm run restart:dfx && npm run build:prod && dfx deploy --network ic",
```
Since dfx 0.9(?) this flag is implied and not necessary anymore.

File: motoko/hello-angular-motoko/karma.conf.js (hunk 7 above)
Comment (sesi200) [reply motoko/hello-angular-motoko/karma.conf.js:2]: yep, that's it

File: motoko/hello-angular-motoko/README.md (hunk 9 above)
Comment (sesi200) [reply motoko/hello-angular-motoko/README.md:45]: I didn't explain myself enough the first time around, sorry. The reasons I didn't like the link to Kyle's article are:
- It's not the standard way to do it (anymore?). One dfx-command does the trick just as well.
- It links to Kyle's _personal_ website. While he's on the SDK team, the article is not an official (and therefore maintained) resource. By now it contains some outdated info (e.g. the URL to the SDK download), and the NNS interface shown in the screenshots is not accurate anymore.
- The link was combined with a bunch of deviations from the instructions. This makes it really hard to follow because people will have to switch between the two pages all the time.

Why I think the new link is fine:
- It's an official (and therefore maintained) resource.
- If we didn't link there, we'd be duplicating information, which makes maintenance much harder.
- It's literally where we send _everyone_ to start. We'd like to use at least that as common base knowledge and assume this in (almost?) all other examples.
- The requirements section is a _lot_ smaller this way.
- It's more clear that this is not a hard requirement, but only for live deployment. (Yes, this is only tangentially related to the link and more about how it's presented, but wanted to mention it anyways)

I hope that you can accept the new link with this reasoning.

(the question about which canister to deploy to is answered in a different comment)

File: motoko/hello-angular-motoko/README.md (hunk 10 above)
Comment (sesi200) [reply motoko/hello-angular-motoko/README.md:86]: Can you at least point people to where they can find out what happens behind the scenes?

File: motoko/hello-angular-motoko/canister_ids.json (hunk 11 above)
Comment (sesi200) [reply motoko/hello-angular-motoko/canister_ids.json:8]: I think this is actually a case of "that's just how dfx works" :)

If this file is not present (or a canister id is not configured for the selected network), `dfx canister create` (or `dfx deploy`, which is basically `dfx canister create && dfx build && dfx canister install`) writes those IDs for you.

This also requires that you have a wallet configured, which is why I'd like to the quickstart guide which describes how to set up a wallet.

File: motoko/hello-angular-motoko/package.json (hunk 13 above)
Comment (sesi200) [reply motoko/hello-angular-motoko/package.json:35]: `dfx new hello && cd hello && dfx deploy` builds and deploys a front end as well, no?

Comment (shawndotey) [reply motoko/hello-angular-motoko/package.json:35]: Yes, it's uses the `webpack` command in the `build` npm script. I should be able to rework this so that we can use the `dfx deploy` instead of `npm run dfx:deploy` seeing that the `dfx deploy` method is preferred.

File: motoko/hello-angular-motoko/README.md (hunk 10 above)
Comment (shawndotey) [reply motoko/hello-angular-motoko/README.md:86]: I think I just need to rework the related scripts to work with `dfx delpoy` instead of `npm run deploy:dfx`

File: motoko/hello-angular-motoko/karma.conf.js (hunk 7 above)
Comment (sesi200) [reply motoko/hello-angular-motoko/karma.conf.js:2]: exactly, that's it

**Comment by sa-github-api** [comment]
Dear @shawndotey,

In order to potentially merge your code in this open-source repository and therefore proceed with your contribution, we need to have your approval on DFINITY's [CLA](https://github.com/dfinity/cla/blob/main/CLA.md).

If you decide to agree with it, please visit this [issue](https://github.com/dfinity/cla/issues/21) and read the instructions there. Once you have signed it, re-trigger the workflow on this PR to see if your code can be merged.

— The DFINITY Foundation
